http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpFullRequest.java

```diff
@@ -66,9 +71,29 @@
         @Override
         default Builder uri(URI uri) {
             return this.protocol(uri.getScheme())
-                       .host(uri.getHost())
-                       .port(uri.getPort())
-                       .encodedPath(SdkHttpUtils.appendUri(uri.getRawPath(), encodedPath()));
+                    .host(uri.getHost())
+                    .port(uri.getPort())
+                    .encodedPath(SdkHttpUtils.appendUri(uri.getRawPath(), encodedPath()));
+        }
+
+        /**
+         * Convenience method to set the {@link #protocol()}, {@link #host()}, {@link #port()},
+         * {@link #encodedPath()} and keep query parameters if present from a {@link URI} object.
+         *
+         * @param uri URI containing protocol, host, port, path and optional query parameters.
+         * @return This builder for method chaining.
+         */
+        default Builder uriWithQueryParams(URI uri) {
+            Builder builder = this.uri(uri);
+            if (uri.getQuery() != null) {
+                Pattern.compile("\\s*&|%26\\s*")
+                       .splitAsStream(uri.getQuery().trim())
+                       .map(s -> s.split("=|%3D", 2))
```

Comment:
Same here, should only splitting on the literal `=`.

Also seems like this assumes each query parameter is a name-value pair; we should support parameters that don't have a value  as well, like in `https://github.com/aws/aws-sdk-java-v2/pull/2082?foo`